### ISSUE TITLE
RFC: add advanced "csv" syntax for "--net" / "--network"

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#31964]
> **Note:** this design is already partly being discussed on https://github.com/docker/docker/pull/27638, but creating this issue for tracking, and wider discussion.

`docker run` and `docker service create` currently don't support adding a container or service to multiple networks. The reason for this is that there is a number of options that can be set _per network_ (for example, network-scoped aliases through `--network-alias`, a fixed IP-address on a network through `--ip` / `--ip6`).

The current situation severely limits networking options for services, as it's currently not possible to;

- Attach/detach services to a network
- Modify networking configuration for a service (`docker service update`) after a service is created
- Set network-aliases for services through the CLI (the API supports this, as does `docker stack deploy`)

The advanced "csv" syntax (as is used for `docker service create --mount`) solves this limitation by grouping options in a single flag, allowing options to be set _per mount_. The same can be done for networks.

### Proposed change

I want to propose adding the same syntax to `--network` and solve this limitation. The advanced syntax should support all network-related options that can be set currently on `docker run` and `docker network connect`, for example;

```bash
docker service create
  --network name=my-network,alias=foo,alias=foo.bar.com,ip=172.30.100.104,ip6=2001:db8::33
```

Suggested options;

Name | Type | Corresponds with | Description
--- | --- | --- | ---
`name` | `string` | `--network` | Name of the network to attach to
`alias` | `string[]` | `--network-alias` | Network-scoped alias for the container / service
`ip` | `string` | `--ip` | IPv4 address for the container / service
`ip6` | `string` | `--ip6` | IPv6 address for the container / service
`link` | `string` | `--link` | Container/service-scoped alias for another container/service (t.b.d.)
`link-local-ip`| `string[]` | `--link-local-ip` | Link-local address(es) for the container

In addition, advanced networking options (as are proposed in https://github.com/docker/docker/pull/27638) can be passed, for example;

- `ipam-opt` (options for the IPAM plugin/driver)
- `opt` (driver-specific options for the network)

Possibly this would also allow options, such as bandwidth limitation per-network for a container/service (see https://github.com/docker/docker/pull/20794, https://github.com/docker/docker/issues/27809, https://github.com/docker/docker/issues/26767, https://github.com/docker/docker/pull/27846)

### To be discussed - per-task/container options

For _services_, some options can be ambiguous, for example, an _alias_, or IP-address can apply to a service as a _whole_, or _per task_ / container. Some thoughts if we want options to be applied to individual tasks;

- provide separate options for attributes that apply to individual containers/tasks (e.g. `container-ip`, `container-alias`)
- support templating (`container-alias={{.Service}}.{{.Slot}}.foobar`)
- support numerators in templating? (`container-ip=172.30.100.x`) - haven't given this one much thought, perhaps that's an IPAM option.

### To be discussed - auto-creation

Do we want the `--network` flag to _create_ networks if they don't exist, or keep the current behavior, and require the network to be created up front?

Personally, I think requiring a network to exist is a cleaner approach (better separation of concerns).

Related issues;

-  [epic] add more options to `service create` / `service update` (https://github.com/docker/docker/issues/25303)
- [UX] Guidelines for short/long syntax flags (https://github.com/docker/docker/issues/28527)
- `--network-alias` option missing on docker service create https://github.com/docker/docker/issues/24787
- multiple networks on `docker run` (https://github.com/docker/docker/issues/17750, https://github.com/docker/docker/issues/17289, https://github.com/docker/docker/pull/17796)
- Add "--network-opt" option as a way to pass arbitrary options to network drivers (https://github.com/docker/docker/pull/27638)

ping @docker/core-engine-maintainers @docker/core-libnetwork-maintainers

**Comment by mavenugo** [comment]
👍 on the idea.

The discussion on task specific options and templating to go with it requires more discussions. There are cases where even such an option may be limiting especially when the task specific option is completely localized to the node where it is being scheduled.

So, in order to keep this proposal manageable , let us try and address the existing limitation and options and accomodate it with the new CSV format.

**Comment by yuyi** [comment]
LGTM

**Comment by stevvooe** [comment]
@thaJeztah The proposal looks good.

A few things to watch out for:

1. Templating is mostly done in the worker, whereas as network allocation happens in the orchestrator. Some of the template context may be unavailable.
2. It would be good to be able to support updating aliases without replacing the full network. I am not sure how we can support that in this syntax, but we should consider addition.

:+1: :+1: :+1: :guitar:

**Comment by deadbeef84** [comment]
I would love to see the `container-alias` option, especially together with swarm/stacks. My understanding is that the only DNS name available right now is the service name itself. Also, being able to assign multiple aliases would be awesome, for instance to allow accessing a service on a specific node:

- `container-alias={{.Slot}}.{{.Service}}.{{.Node}}`
- `container-alias={{.Service}}.{{.Node}}`

Any chance of this happening?

**Comment by EliSnow** [comment]
#33130 provides name and alias through the CSV syntax. Unfortunately it is undocumented.

I would really like to see `container-alias` with templating support.

**Comment by thaJeztah** [comment]
Yes, https://github.com/moby/moby/pull/33130 / https://github.com/docker/cli/pull/62 adds initial options for services; https://github.com/docker/cli/pull/843 is addressing the missing documentation

**Comment by thaJeztah** [comment]
CLI implementation for `docker run` / `docker create` is worked on in https://github.com/docker/cli/pull/317
